### PR TITLE
Suggest to users: RMM gives 50% speedboost to cudf pathway for `map_index`

### PR DESCRIPTION
**Pull request by puririshi98**
repro:
cd /opt/pyg; pip uninstall -y torch-geometric; rm -rf pytorch_geometric; git clone -b map_index_rmm https://github.com/pyg-team/pytorch_geometric.git; cd /opt/pyg/pytorch_geometric; pip install .; python3 test/utils/test_map.py 

w/ RMM:
```
root@79c6dc981596:/opt/pyg/pytorch_geometric# python3 test/utils/test_map.py 
Inclusive:
+-----------+-----------+
| Name      | Forward   |
|-----------+-----------|
| map_index | 0.1987s   |
+-----------+-----------+
Exclusive:
+-----------+-----------+
| Name      | Forward   |
|-----------+-----------|
| map_index | 0.2195s   |
```
+-----------+-----------+
w/o:
```
root@79c6dc981596:/opt/pyg/pytorch_geometric# python3 test/utils/test_map.py 
Inclusive:
+-----------+-----------+
| Name      | Forward   |
|-----------+-----------|
| map_index | 0.3201s   |
+-----------+-----------+
Exclusive:
+-----------+-----------+
| Name      | Forward   |
|-----------+-----------|
| map_index | 0.3158s   |
+-----------+-----------+
```

### CODE REVIEW

**Comment by puririshi98** [comment]
from an nvidia devtech
> My two cents is that switching the allocator slows down PyTorch operations and also increases fragmentation chances .

**Comment by wence-** [comment]
Just to provide a bit more detail to that comment. The proposal in this PR (to switch up the allocator in this one place) would mean that there might potentially be more than one pool allocator fighting for space. The default RMM pool is a relatively straightforward one which, for various reasons around cuda alignment requirements, can easily lead to fragmentation if the application makes lots of small allocations.

That is not to say that it is full-stop a bad idea. However, one needs a certain amount of coordination amongst different libraries when using pool allocators. RMM provides allocator callbacks for torch (as seen here), cupy, and numba. And really the user is the only one who knows which set of packages they are using (and hence which allocators want configured).

So, it can be worthwhile, but it is better to do so as an explicit user choice.

**Comment by puririshi98** [comment]
@rusty1s plz adjust doc changes as desired

File: torch_geometric/utils/map.py [hunk 1]
```diff
@@ -40,6 +40,18 @@ def map_index(
 
         >>> map_index(src, index)
         (tensor([1, 2, 2, 0]), tensor([True, True, False, True, True]))
+
+    Note:
```
Comment (rusty1s) [inline torch_geometric/utils/map.py:44]:
```suggestion
    .. note::
```

File: torch_geometric/utils/map.py [hunk 2]
```diff
@@ -40,6 +40,18 @@ def map_index(
...
+
+        If src is a cuda tensor, consider using RMM for significant speed boosts.
+        Proceed with caution as RMM may conflict with other allocators or fragment.
+
+    .. code-block::
```
Comment (rusty1s) [inline torch_geometric/utils/map.py:49]:
```suggestion
    .. code-block:: python
```

File: torch_geometric/utils/map.py [hunk 3]
```diff
@@ -40,6 +40,18 @@ def map_index(
 
         >>> map_index(src, index)
         (tensor([1, 2, 2, 0]), tensor([True, True, False, True, True]))
+
+    Note:
+
+        If src is a cuda tensor, consider using RMM for significant speed boosts.
```
Comment (rusty1s) [inline torch_geometric/utils/map.py:46]:
```suggestion
        If inputs are on GPU, consider using RMM for significant speed boosts.
```

**Review by rusty1s — APPROVED** [review]

File: torch_geometric/utils/map.py [hunk 4]
```diff
@@ -40,6 +40,18 @@ def map_index(
 
         >>> map_index(src, index)
         (tensor([1, 2, 2, 0]), tensor([True, True, False, True, True]))
+
+    .. note::
+
+        If inputs are on GPU, consider using RMM for significant speed boosts.
+        Proceed with caution as RMM may conflict with other allocators or fragment.
```
Comment (rusty1s) [inline torch_geometric/utils/map.py:47]:
```suggestion
        Proceed with caution as RMM may conflict with other allocators or fragments.
```

File: torch_geometric/utils/map.py [hunk 5]
```diff
@@ -40,6 +40,18 @@ def map_index(
...
+
+    .. note::
+
+        If inputs are on GPU, consider using RMM for significant speed boosts.
+        Proceed with caution as RMM may conflict with other allocators or fragments.
+
+    .. code-block:: python
+
+        import rmm
+        rmm.reinitialize(pool_allocator=True)\n \
```
Comment (rusty1s) [inline torch_geometric/utils/map.py:52]:
```suggestion
        rmm.reinitialize(pool_allocator=True)
```

File: torch_geometric/utils/map.py [hunk 6]
```diff
@@ -40,6 +40,18 @@ def map_index(
...
+    .. note::
+
+        If inputs are on GPU, consider using RMM for significant speed boosts.
+        Proceed with caution as RMM may conflict with other allocators or fragments.
+
+    .. code-block:: python
+
+        import rmm
+        rmm.reinitialize(pool_allocator=True)
+        torch.cuda.memory.change_current_allocator(rmm.rmm_torch_allocator)"
```
Comment (rusty1s) [inline torch_geometric/utils/map.py:53]:
```suggestion
        torch.cuda.memory.change_current_allocator(rmm.rmm_torch_allocator)
```

**Pull request merged**